presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java

```diff
@@ -299,23 +299,28 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
         // NOTE: buffers are allowed to be created in the FINISHED state because destroy() can move to the finished state
         // without a clean "no-more-buffers" message from the scheduler.  This happens with limit queries and is ok because
         // the buffer will be immediately destroyed.
-        checkState(state.get().canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
+        BufferState state = this.state.get();
+        checkState(state.canAddBuffers() || !outputBuffers.isNoMoreBufferIds(), "No more buffers already set");
 
         // NOTE: buffers are allowed to be created before they are explicitly declared by setOutputBuffers
         // When no-more-buffers is set, we verify that all created buffers have been declared
         buffer = new ClientBuffer(taskInstanceId, id);
 
-        // add initial pages
-        buffer.enqueuePages(initialPagesForNewBuffers);
+        // do not setup the new buffer if we are already failed
+        if (state != FAILED) {
+            // add initial pages
+            buffer.enqueuePages(initialPagesForNewBuffers);
 
-        // update state
-        if (!state.get().canAddPages()) {
-            buffer.setNoMorePages();
-        }
+            // update state
+            if (!state.canAddPages()) {
+                // DO NOT set no more pages if FAILED
```

Comment:
This comment is a little confusing, since state is guaranteed to be NOT FAILED at this point. I'd rephrase as "set no more pages only if not FAILED" and maybe add a little more context.